updating pycubed firmware for mainboardv05

maholli opened the pull request:
PyCubed users: 

1. Note pin changes between mainboardv04 to mainboardv05!
2. Remember that pycubed's low-level boot counter is not present on the circuitpython main branch. This is because a truly low-level boot counter implementation requires modifications to circuitpython's main.c file.
    - To implement the PyCubed boot counter for yourself, add the following to `/main.c`
       ```cpp
       #include "shared-bindings/nvm/ByteArray.h" // PYCUBED BOOT COUNTER
       ```
       Prior to the `run_code_py()` definition, add:
       ```cpp
       // PYCUBED BOOT COUNTER
       nvm_bytearray_obj_t bootcnt = {
           .base = {
               .type = &nvm_bytearray_type
                   },
           .len = ( uint32_t) 8192,
           .start_address = (uint8_t*) (0x00080000 - 8192)
           };
       uint8_t value_out = 0;
       ```
       Then INSIDE the `run_code_py()` function definition, and before the line `pyexec_result_t result;`, add:
       ```cpp
           // PYCUBED BOOT COUNTER
           common_hal_nvm_bytearray_get_bytes(&bootcnt,0,1,&value_out);
           ++value_out;
           common_hal_nvm_bytearray_set_bytes(&bootcnt,0,&value_out,1);
       ```

tannewt commented:
How about making a new board def for the new revision? That way the old one will still work. You can add the boot counter behind a flag and then merge it in if you want to be on main.

maholli commented:
> 
> 
> How about making a new board def for the new revision? That way the old one will still work. You can add the boot counter behind a flag and then merge it in if you want to be on main.

@tannewt are seperate board defs your preferred method of handling major board revisions? Part of me yearns for a cleaner solution than that, especially since folks are wanting firmware for multiple D51 flash sizes (J19 and J20, for example). Maybe solution is discrete board defs but easily managed via `revisions` or `configuration` drop down on the board's circuitpython.org/downloads page?

Excited about having the boot counter feature on main. Will do! Thanks!

tannewt commented:
Ya, separate board defs are preferred. Board revisions in my book don't change pinouts. Revisions would only change layout or compatible parts.

We have lots of infrastructure to handle many boards. (I just added a tweak to CI to only build changed boards for PRs.) Adding another level like revision or configuration will just add complexity to the tooling.

maholli commented:
@tannewt if it's alright with you, I've elected not to create separate board defs for the PyCubed update since it would require 6 new defs in the boards directory: pycubed-v04_J19, pycubed-v04_J20, pycubed-v05_J19, pycubed-v05_J20.

I did, however, add the boot counter in main.c. When you have a moment, can you review the implementation? Usage details below.

## Boot Counter Usage

The boot counter is a uint8_t single-byte counter stored in the first NVM byte position (`micrcontroller.nvm[0]`). The counter increments by 1 each time the board boots, regardless if it's a hard or soft reset. 

Enable the boot counter by adding `#define CIRCUITPY_BOOT_COUNTER 1` to your board's mpconfigboard.h file. Note that an error will be thrown during the build if `CIRCUITPY_INTERNAL_NVM_SIZE` is not also set within mpconfigboard.h.

tannewt commented on main.c:
```diff
@@ -304,6 +308,15 @@ STATIC void print_code_py_status_message(safe_mode_t safe_mode) {
 }
 
+#ifdef CIRCUITPY_BOOT_COUNTER
+nvm_bytearray_obj_t bootcnt = {
+    .base = {.type = &nvm_bytearray_type},
+    .len = ( uint32_t) 1,
+    .start_address = (uint8_t*) (0x00080000 - CIRCUITPY_INTERNAL_NVM_SIZE)
+    };
```
Why do you need this? Can't you use the nvm object instead?

maholli replied:
I was never able to get `common_hal_nvm_bytearray_get_bytes(&common_hal_mcu_nvm_obj,0,1,&value_out);` to work since `common_hal_mcu_nvm_obj` is declared as a const. Any thoughts?

tannewt replied:
I think you can add the `const` to the common_hal_nvm_bytearray functions and then it will.

tannewt reviewed (changes requested):
I really don't like you changing the original board. It's a bad policy because the old boards can never be updated then. Just add two new boards v5 and v5 mram.

maholli commented on shared-bindings/nvm/ByteArray.h:
```diff
@@ -31,13 +31,13 @@
 extern const mp_obj_type_t nvm_bytearray_type;
 
-uint32_t common_hal_nvm_bytearray_get_length(nvm_bytearray_obj_t *self);
+uint32_t common_hal_nvm_bytearray_get_length(const nvm_bytearray_obj_t *self);
```
@tannewt specifying `const` for the nvm object builds and works in my cursory tests, allowing me to use `common_hal_mcu_nvm_obj` rather than creating my own. Just not sure if this has longer reaching implications?

maholli reviewed (commented):
I think this fixes nvm object concerns, would be nice to have a sanity check of the implications, though.

dhalbert commented:
@maholli are the boards distributed, or are these only in your lab? I can see you not wanting a Cartesian product of new boards for every revision.

maholli commented:
> 
> 
> @maholli are the boards distributed, or are these only in your lab? I can see you not wanting a Cartesian product of new boards for every revision.

Good question. Yes, students have built hundreds of these boards around the world, but I also expect the same to occur for the new revision. Part of me feels folks can just dig up the appropriate commit if they want to build CP for a prior revision (with minor fiddling), but I also appreciate Scott's reasoning. I'll stick to new board defs for now.

maholli commented:
Ok @tannewt, with https://github.com/adafruit/circuitpython/pull/5302/commits/dbd557dc89406387003967837f1a2b17d53c4e31 this PR has changed the following:

1. Split existing board defs `pycubed` and `pycubed_mram` into:
    a. `pycubed_v04` and `pycubed_mram_v04`
    b. `pycubed_v05` and `pycubed_mram_v05`
    An update to circuitpython-org repo will be merged once this PR is accepted.
2. Added boot counter to main.c using existing `common_hal_mcu_nvm_obj` object. 

Ready for checking and approval.

tannewt commented on main.c:
```diff
@@ -316,6 +321,12 @@ STATIC bool run_code_py(safe_mode_t safe_mode) {
     }
     #endif
 
+    #ifdef CIRCUITPY_BOOT_COUNTER
```
This isn't really a boot counter because you are incrementing every code.py run. How about putting it in `board_init()` instead? It's a small amount of code so I think it's ok to duplicate it in each board's `board_init`.

maholli replied:
@tannewt putting the counter in `board_init()` or `reset_board()` resulted in the counter being incremented by 2 sometimes 3 per reboot/restart/power cycle. I figured this was a result of various VM behavior?

dhalbert replied:
`board_init()` is called after fake sleep, as well as once in `main()`. `reset_board()` is called in `main()` and `cleanup_after_vm()`, so would be called twice (once for possible `boot.py` and once for `code.py`).

You could add your own routine just once to `main()` and conditionalize it with `CIRCUITPY_BOOT_COUNTER`, or something like that.

maholli replied:
Thanks @dhalbert! I'll move the `#ifdef CIRCUITPY_BOOT_COUNTER` to `main()`

dhalbert replied:
Yes, let's see if it works. BTW, use `#if` instead of `#ifdef`, and set it to `0` as the default. That is safer all around. We use that style for all `CIRCUITPY_*` flags.

tannewt commented on ports/atmel-samd/boards/pycubed_mram_v04/pins.c:
```diff
@@ -1,13 +1,12 @@
     { MP_ROM_QSTR(MP_QSTR_MOSI),     MP_ROM_PTR(&pin_PA12)  },
     { MP_ROM_QSTR(MP_QSTR_MISO),     MP_ROM_PTR(&pin_PA14)  },
-    { MP_ROM_QSTR(MP_QSTR_xSDCS),    MP_ROM_PTR(&pin_PA27)  },
+    { MP_ROM_QSTR(MP_QSTR_SD_CS),    MP_ROM_PTR(&pin_PA27)  },
```
Generally we only add newer names and don't remove them. Removing them breaks existing code.

tannewt reviewed (changes requested):
Please don't change the id (aka folder name) of the existing boards. We've done it in rare cases when something hasn't been released yet. In this case, since it is released, you'll create a discontinuity with all folks currently using it. You can add a comment to one of the files that it is V4 of the board. (One discontinuity would not being able to see all of the older releases like you can here: https://adafruit-circuit-python.s3.amazonaws.com/index.html?prefix=bin/pycubed/en_US/)

maholli commented:
> 
> 
> Please don't change the id (aka folder name) of the existing boards. We've done it in rare cases when something hasn't been released yet. In this case, since it is released, you'll create a discontinuity with all folks currently using it. You can add a comment to one of the files that it is V4 of the board. (One discontinuity would not being able to see all of the older releases like you can here: https://adafruit-circuit-python.s3.amazonaws.com/index.html?prefix=bin/pycubed/en_US/)

Completely understood. I restored the pycubed board defs. @tannewt Do you prefer the redundant (but labeled) pycubed_v04 board def be removed?

tannewt commented:
> Completely understood. I restored the pycubed board defs. @tannewt Do you prefer the redundant (but labeled) pycubed_v04 board def be removed?

Yes, please remove the new v4 def and leave the existing one.

maholli commented:
> 
> 
> > Completely understood. I restored the pycubed board defs. @tannewt Do you prefer the redundant (but labeled) pycubed_v04 board def be removed?
> 
> Yes, please remove the new v4 def and leave the existing one.

Done. 👍

dhalbert commented:
Each commit causes a set of builds, so I am cancelling the older ones just to save build time.

tannewt commented on tools/ci_check_duplicate_usb_vid_pid.py:
```diff
@@ -35,6 +35,10 @@
     "circuitplayground_express_displayio",
     "pycubed",
     "pycubed_mram",
+    "pycubed_v04",
+    "pycubed_v05",
+    "pycubed_mram_v04",
+    "pycubed_mram_v05",
```
```suggestion
    "pycubed_v05",
    "pycubed_mram_v05",
```

tannewt commented on main.c:
```diff
@@ -107,6 +107,11 @@
 #include "shared-bindings/wifi/__init__.h"
 #endif
 
+#ifdef CIRCUITPY_BOOT_COUNTER
```
In `py/circuitpy_mpconfig.h` put an `ifndef` that defines this as 0. That way we ensure that the correct files were included to have this defined.

```suggestion
#if CIRCUITPY_BOOT_COUNTER
```

tannewt commented on main.c:
```diff
@@ -796,6 +801,13 @@ int __attribute__((used)) main(void) {
     // Turn on RX and TX LEDs if we have them.
     init_rxtx_leds();
 
+    #ifdef CIRCUITPY_BOOT_COUNTER
```
```suggestion
    #if CIRCUITPY_BOOT_COUNTER
```

tannewt reviewed (changes requested):
Two more minor things. Thanks!
